### ISSUE TITLE
ClassNotFound exception when running figwheel and :modules configuration is present in project.clj in another compilation target.

### ISSUE THREAD

**Issue by Gonzih** [bhauman/lein-figwheel#104]
Hello,

I'm running figwheel in develompent and I have dev compilation target without any optimizations enabled (:none). Works like a charm! But also I have anonther production specific compilation target with advanced compilation mode and once I add :modules configuration option to this production configuration target figwheel starts to fail in development. Here is exception stacktrace:

```
Exception in thread "main" java.lang.ClassNotFoundException: myapp.page-two.core, compiling:(/tmp/form-init2027442236028884163.clj:1:555)
	at clojure.lang.Compiler.analyze(Compiler.java:6464)
	at clojure.lang.Compiler.analyze(Compiler.java:6406)
	at clojure.lang.Compiler$SetExpr.parse(Compiler.java:3064)
	at clojure.lang.Compiler.analyze(Compiler.java:6455)
	at clojure.lang.Compiler.analyze(Compiler.java:6406)
	at clojure.lang.Compiler$MapExpr.parse(Compiler.java:2981)
	at clojure.lang.Compiler.analyze(Compiler.java:6453)
	at clojure.lang.Compiler.analyze(Compiler.java:6406)
	at clojure.lang.Compiler$MapExpr.parse(Compiler.java:2981)
	at clojure.lang.Compiler.analyze(Compiler.java:6453)
	at clojure.lang.Compiler.analyze(Compiler.java:6406)
	at clojure.lang.Compiler$MapExpr.parse(Compiler.java:2981)
	at clojure.lang.Compiler.analyze(Compiler.java:6453)
	at clojure.lang.Compiler.analyze(Compiler.java:6406)
	at clojure.lang.Compiler$MapExpr.parse(Compiler.java:2981)
	at clojure.lang.Compiler.analyze(Compiler.java:6453)
	at clojure.lang.Compiler.analyze(Compiler.java:6406)
	at clojure.lang.Compiler$VectorExpr.parse(Compiler.java:3126)
	at clojure.lang.Compiler.analyze(Compiler.java:6447)
	at clojure.lang.Compiler.analyze(Compiler.java:6406)
	at clojure.lang.Compiler$MapExpr.parse(Compiler.java:2981)
	at clojure.lang.Compiler.analyze(Compiler.java:6453)
	at clojure.lang.Compiler.analyze(Compiler.java:6406)
	at clojure.lang.Compiler$InvokeExpr.parse(Compiler.java:3719)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6646)
	at clojure.lang.Compiler.analyze(Compiler.java:6445)
	at clojure.lang.Compiler.analyze(Compiler.java:6406)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:5782)
	at clojure.lang.Compiler$LetFnExpr$Parser.parse(Compiler.java:5916)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6644)
	at clojure.lang.Compiler.analyze(Compiler.java:6445)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6632)
	at clojure.lang.Compiler.analyze(Compiler.java:6445)
	at clojure.lang.Compiler.analyze(Compiler.java:6406)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:5782)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6644)
	at clojure.lang.Compiler.analyze(Compiler.java:6445)
	at clojure.lang.Compiler.analyze(Compiler.java:6406)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:5782)
	at clojure.lang.Compiler$TryExpr$Parser.parse(Compiler.java:2191)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6644)
	at clojure.lang.Compiler.analyze(Compiler.java:6445)
	at clojure.lang.Compiler.analyze(Compiler.java:6406)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:5782)
	at clojure.lang.Compiler$FnMethod.parse(Compiler.java:5217)
	at clojure.lang.Compiler$FnExpr.parse(Compiler.java:3846)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6642)
	at clojure.lang.Compiler.analyze(Compiler.java:6445)
	at clojure.lang.Compiler.eval(Compiler.java:6700)
	at clojure.lang.Compiler.eval(Compiler.java:6693)
	at clojure.lang.Compiler.load(Compiler.java:7130)
	at clojure.lang.Compiler.loadFile(Compiler.java:7086)
	at clojure.main$load_script.invoke(main.clj:274)
	at clojure.main$init_opt.invoke(main.clj:279)
	at clojure.main$initialize.invoke(main.clj:307)
	at clojure.main$null_opt.invoke(main.clj:342)
	at clojure.main$main.doInvoke(main.clj:420)
	at clojure.lang.RestFn.invoke(RestFn.java:421)
	at clojure.lang.Var.invoke(Var.java:383)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.Var.applyTo(Var.java:700)
	at clojure.main.main(main.java:37)
Caused by: java.lang.ClassNotFoundException: myapp.page-two.core
	at java.net.URLClassLoader$1.run(URLClassLoader.java:372)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:360)
	at clojure.lang.DynamicClassLoader.findClass(DynamicClassLoader.java:61)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:344)
	at clojure.lang.RT.classForName(RT.java:2065)
	at clojure.lang.Compiler.resolveIn(Compiler.java:6921)
	at clojure.lang.Compiler.resolve(Compiler.java:6884)
	at clojure.lang.Compiler.analyzeSymbol(Compiler.java:6845)
	at clojure.lang.Compiler.analyze(Compiler.java:6427)
	... 61 more
```

My directories structure looks like that:

```
src/
└── myapp
    ├── core.cljs
    ├── page_one
    │   ├── core.cljs
    ├── page_two
    │   └── core.cljs
dev_src/
└── myapp
    └── dev.cljs
```

and here is my dev ns:

```clojure
(ns ^:figwheel-always myapp.dev
  (:require
            [myapp.page-one.core]
            [myapp.page-two.core]
            [figwheel.client :as fw :include-macros true]))

(fw/watch-and-reload
  :websocket-url   "ws://localhost:3449/figwheel-ws"
  :jsload-callback (fn [] (print "reloaded")))
```

Commenting out :modules configuration solves this issue in development. I have no idea where to look further, any tips are welcome! Thanks!

**Comment by bhauman** [comment]
Does `lein cljsbuild auto` work for your dev build?

paging @swannodette

**Comment by cichli** [comment]
Just to confirm - you're running Figwheel on your `:optimizations :none` build and it breaks when you configure modules for another build, or you're trying to run Figwheel on an `:optimizations :advanced` build? If the latter: Figwheel only supports `:optimizations :none`.

**Comment by swannodette** [comment]
People need to stop reporting errors like this without taking the time to discover the root cause (I am saying this in a nice voice). The very first thing anyone should do is determine whether the problem is in ClojureScript itself before reporting to figwheel or cljsbuild or whatever. I admit it's been difficult to test the ClojureScript compiler quickly and directly - this is being fixed *right now* and moving forward I really don't want to be involved in downstream bug reports :) I think it will make downstream tool maintainers happier (since they don't have to sift through fundamental issues) and I can maintain focus on the fundamentals. Thanks!

**Comment by Gonzih** [comment]
yes `lein cljsbuild auto` dev works fine.
figwheel is running for `:optimizations :none`. modules configuration is in different compilation target and this modules configuration affects dev compilation target.

@swannodette that is why this issue is not just bug report but also attempt to get help on how to debug this problem.

this issue is clearly related to figwheel because cljsbuild works like a charm for all builds in my code.
also it's really hard to understand what is going on (except that something goes wrong in clojure.lang.Compiler.analyze code) based o given stack trace.

**Comment by Gonzih** [comment]
Here is how my :cljsbuild configuration looks like:

```clojure
  :cljsbuild {:builds [{:id "dev"
                        :source-paths ["src" "dev_src"]
                        :compiler {:output-to "resources/public/javascripts/vd.js"
                                   :source-map true
                                   :source-map-timestamp true
                                   :output-dir "resources/public/javascripts/out"
                                   :asset-path "http://localhost:3449/javascripts/out"
                                   :optimizations :none
                                   :warnings true
                                   :cache-analysis true
                                   :main myapp.dev}}
                       {:id "production"
                        :source-paths ["src"]
                        :compiler {:output-to "public/javascripts/myapp.js"
                                   :source-map "public/javascripts/myapp.js.map"
                                   :output-dir "public/javascripts/out"
                                   :preamble [...]
                                   ; In case of error please comment this out
                                   :modules {:cljs-base {:output-to "public/javascripts/base.js"}
                                             :page-one {:output-to "public/javascripts/page-one.js"
                                                            :entries #{myapp.page-one.core}}
                                             :page-two {:output-to "public/javascripts/page-two.js"
                                                         :entries #{myapp.page-two.core}}}
                                   :externs [...]
                                   :pretty-print false
                                   :cache-analysis true
                                   :warnings true
                                   :optimizations :advanced}}]}
```

**Comment by bhauman** [comment]
Actually i know what it is.
The bare symbols in your module config are crossing the serialization boundary from lein to the eval-in-project  environment.  Just use strings in your module declarations instead or bare symbols for now.  

i.e. "myapp.page-two.core"

This has been a continual pain in my butt. I should just do a postwalk and replace all symbols in the config.

**Comment by Gonzih** [comment]
@bhauman you are totally right, it fixed the issue, thank you
@swannodette I updated clojurescript wiki page https://github.com/clojure/clojurescript/wiki/Compiler-Options#modules

**Comment by Gonzih** [comment]
or should I revert update in wiki then? is it issue in figwheel then?

**Comment by bhauman** [comment]
This is still an issue I am going to solve.

**Comment by bhauman** [comment]
yeah revert the update.

**Comment by Gonzih** [comment]
reverted

**Comment by bhauman** [comment]
Just pushed a fix for this. Haven't deployed it yet.

**Comment by Gonzih** [comment]
Thank you very much!

**Comment by bhauman** [comment]
Just released 0.2.5 with this fixed.
